state/unit.go

```diff
@@ -284,6 +294,65 @@ func (u *Unit) PasswordValid(password string) bool {
 	return false
 }
 
+// UpdateOperation returns a model operation that will update a unit.
+func (u *Unit) UpdateOperation(props UnitUpdateProperties) *UpdateUnitOperation {
+	return &UpdateUnitOperation{
+		unit:  &Unit{st: u.st, doc: u.doc},
+		props: props,
+	}
+}
+
+// UpdateUnitsOperation is a model operation for updating a unit.
+type UpdateUnitOperation struct {
+	unit  *Unit
+	props UnitUpdateProperties
+
+	setStatusDocs map[string]statusDoc
+}
+
+// Build is part of the ModelOperation interface.
+func (op *UpdateUnitOperation) Build(attempt int) ([]txn.Op, error) {
+	// TODO(caas) - we don't yet model containers for units so do not update address or port
+
+	op.setStatusDocs = make(map[string]statusDoc)
+	ops := []txn.Op{{
+		C:      unitsC,
+		Id:     op.unit.doc.DocID,
+		Assert: txn.DocExists,
+		Update: bson.D{{"$set", bson.D{{"provider-id", op.props.ProviderId}}}},
```

Comment:
AFAICT, this doesn't make sense.

In the code where you construct the operation, you're matching the units by the provider ID. So props.ProviderId should be the same as the unit's current provider-id field value. I think the provider ID should be considered immutable?

Reply:
It is immutable. But this code is for the case where a unit doc is first created as a result of add-unit. It doesn't yet have a provider-id set. Then the provider sends some unit records and one of those is used to update this previously "unallocated" juju unit doc. The next time an update of that cloud unit occurs, this is then a no-op.

Reply:
Sorry, I missed that it was being used in two different ways. In that case, I'd suggest checking if props.Provider != "", and only then setting; and asserting that the current value is ""